File: archive/cdk-ts/Pr789/lib/security-group-stack.ts
import * as ec2 from 'aws-cdk-lib/aws-ec2';
import { Construct } from 'constructs';

interface SecurityGroupStackProps {
  vpc: ec2.IVpc;
  environmentSuffix?: string;
}

export class SecurityGroupStack extends Construct {
  public readonly webSecurityGroup: ec2.SecurityGroup;
  public readonly sshSecurityGroup: ec2.SecurityGroup;

  constructor(scope: Construct, id: string, props: SecurityGroupStackProps) {
    super(scope, id);

    const environmentSuffix = props?.environmentSuffix || 'dev';

    // Security Group for Web servers (HTTP traffic)
    this.webSecurityGroup = new ec2.SecurityGroup(
      this,
      `WebSecurityGroup-${environmentSuffix}`,
      {
        vpc: props.vpc,
        description: 'Security group for web servers allowing HTTP traffic',
        allowAllOutbound: true,
      }
    );

    // Allow HTTP traffic from anywhere
    this.webSecurityGroup.addIngressRule(
      ec2.Peer.anyIpv4(),
      ec2.Port.tcp(80),
      'Allow HTTP traffic'
    );

    // Allow HTTPS traffic from anywhere
    this.webSecurityGroup.addIngressRule(
      ec2.Peer.anyIpv4(),
      ec2.Port.tcp(443),
      'Allow HTTPS traffic'
    );

    // Security Group for SSH access
    this.sshSecurityGroup = new ec2.SecurityGroup(
      this,
      `SshSecurityGroup-${environmentSuffix}`,
      {
        vpc: props.vpc,
        description: 'Security group for SSH access',
        allowAllOutbound: true,
      }
    );

    // Allow SSH traffic from anywhere (in production, restrict to specific IPs)
    this.sshSecurityGroup.addIngressRule(
      ec2.Peer.anyIpv4(),
      ec2.Port.tcp(22),
      'Allow SSH traffic'
    );

    // Security Group for internal communication
    const internalSecurityGroup = new ec2.SecurityGroup(
      this,
      `InternalSecurityGroup-${environmentSuffix}`,
      {
        vpc: props.vpc,
        description: 'Security group for internal VPC communication',
        allowAllOutbound: true,
      }
    );

    // Allow all traffic from VPC CIDR
    internalSecurityGroup.addIngressRule(
      ec2.Peer.ipv4('10.0.0.0/16'),
      ec2.Port.allTraffic(),
      'Allow internal VPC traffic'
    );
  }
}

File: archive/cdk-java/Pr2283/lib/src/main/java/app/SecurityGroupStack.java
package app;

import software.amazon.awscdk.Stack;
import software.amazon.awscdk.StackProps;
import software.amazon.awscdk.services.ec2.*;
import software.constructs.Construct;

import java.util.Map;

public class SecurityGroupStack extends Stack {

    private final SecurityGroup webSecurityGroup;
    private final SecurityGroup dbSecurityGroup;

    public SecurityGroupStack(final Construct scope, final String id, 
                            final StackProps props, final Vpc vpc) {
        super(scope, id, props);

        // Security Group for EC2 web server
        this.webSecurityGroup = SecurityGroup.Builder.create(this, "app-sg-web")
                .securityGroupName("app-sg-web")
                .description("Security group for web servers - HTTPS only from specific ranges")
                .vpc(vpc)
                .build();

        // Allow HTTPS from specific IP ranges only
        webSecurityGroup.addIngressRule(
                Peer.ipv4("10.0.0.0/16"),
                Port.tcp(443),
                "HTTPS from internal network"
        );
        webSecurityGroup.addIngressRule(
                Peer.ipv4("192.168.1.0/24"),
                Port.tcp(443),
                "HTTPS from management network"
        );

        // Security Group for RDS
        this.dbSecurityGroup = SecurityGroup.Builder.create(this, "app-sg-database")
                .securityGroupName("app-sg-database")
                .description("Security group for RDS database")
                .vpc(vpc)
                .build();

        // Allow MySQL/Aurora connections from web security group only
        dbSecurityGroup.addIngressRule(
                Peer.securityGroupId(webSecurityGroup.getSecurityGroupId()),
                Port.tcp(3306),
                "MySQL from web servers"
        );

        this.addCommonTags();
    }

    private void addCommonTags() {
        Map<String, String> tags = Map.of(
                "Project", "CloudSecurity",
                "Environment", "Production"
        );
        tags.forEach((key, value) -> this.getNode().addMetadata(key, value));
    }

    public SecurityGroup getWebSecurityGroup() {
        return webSecurityGroup;
    }

    public SecurityGroup getDbSecurityGroup() {
        return dbSecurityGroup;
    }
}